### PR TITLE
fix(test): exclude virtual atoms from force and Hessian metrics

### LINKED ISSUE #5928: [BUG] dp test includes type == -1 virtual atoms in force and Hessian RMSE denominators
### Bug summary

`dp test` recognizes `type == -1` atoms as virtual/padding atoms in mixed-type datasets, but its force and Hessian error metrics are computed over the complete padded arrays without applying an atom mask.

As a result, entries associated with virtual atoms participate in the RMSE denominator. Even when their prediction and reference errors are both zero, they increase the denominator and artificially lower the reported RMSE. If those entries are nonzero, they also affect the numerator.

### DeePMD-kit version

- Source inspected: current `master` at commit [721f586](https://github.com/deepmodeling/deepmd-kit/commit/721f586b36b107c6662e5efe5d5b6886faebf386)
- Locally installed `dp --version`: `DeePMD-kit v3.2.0b1.dev105+gf23d383ef.d20260705`

### Backend and installation

This appears backend-independent because the behavior is in the shared `dp test` entry point and evaluation-metric helpers. The source tree was checked from the current master branch.

Installation method: built from source.

### Current behavior

The number of atoms is taken from the full padded type-array width:

https://github.com/deepmodeling/deepmd-kit/blob/721f586b36b107c6662e5efe5d5b6886faebf386/deepmd/entrypoints/test.py#L677-L680

Force predictions and references are passed to the common metric calculation without a valid-atom mask:

https://github.com/deepmodeling/deepmd-kit/blob/721f586b36b107c6662e5efe5d5b6886faebf386/deepmd/entrypoints/test.py#L769-L790

https://github.com/deepmodeling/deepmd-kit/blob/721f586b36b107c6662e5efe5d5b6886faebf386/deepmd/utils/eval_metrics.py#L170-L174

Hessian predictions and references are handled in the same way:

https://github.com/deepmodeling/deepmd-kit/blob/721f586b36b107c6662e5efe5d5b6886faebf386/deepmd/entrypoints/test.py#L838-L845

The common metric uses all elements for both `np.mean` and the cross-system aggregation weight `diff.size`:

https://github.com/deepmodeling/deepmd-kit/blob/721f586b36b107c6662e5efe5d5b6886faebf386/deepmd/utils/eval_metrics.py#L46-L48

https://github.com/deepmodeling/deepmd-kit/blob/721f586b36b107c6662e5efe5d5b6886faebf386/deepmd/utils/eval_metrics.py#L114-L126

The data loader already detects and counts `type == -1` atoms, but that information is not used when calculating these metrics:

https://github.com/deepmodeling/deepmd-kit/blob/721f586b36b107c6662e5efe5d5b6886faebf386/deepmd/utils/data.py#L749-L780

### Steps to reproduce

1. Prepare a mixed-type test dataset whose `real_atom_types.npy` contains at least one `-1` virtual/padding atom.
2. Provide force labels and, for a Hessian-enabled model, Hessian labels.
3. Run `dp test`.
4. Compare the reported values against manually masked calculations using `atype >= 0`.

For force, the current denominator is effectively:

```text
nframes * natoms * 3
```

The expected denominator is:

```text
sum(number of atoms with type >= 0 in each frame) * 3
```

For Hessian, a matrix element should be valid only when both of its coordinate indices belong to atoms with `type >= 0`.

### Expected behavior

- Exclude force components belonging to `type == -1` atoms from both the numerator and denominator.
- For Hessian, exclude every element for which either coordinate belongs to a `type == -1` atom.
- Use the number of valid scalar elements as the aggregation weight across systems.
- Add regression tests containing mixed-type frames with different numbers or locations of virtual atoms.

It may also be useful to audit other per-atom metrics and `Energy RMSE/Natoms`, since the current `natoms` value includes the padded `-1` slots.

Coding agent: Codex
Codex version: codex-cli 0.144.6
Model: gpt-5.6-sol
Reasoning effort: xhigh

### PR DESCRIPTION
**Pull request by LOGO127**
## Summary

Fixes #5928.

`dp test` currently reduces force and Hessian errors over padded arrays. Type `-1` slots lower MAE/RMSE even when both values are zero, and nonzero/NaN padding can contaminate the numerator. The padded array size is also reused as the chunk/system aggregation weight.

- Select real-atom force components before computing errors, including optional `atom_pref` weighting.
- Select Hessian entries only when both Cartesian indices belong to real atoms in the same frame.
- Let the selected scalar count (or selected preference sum) drive existing chunk and cross-system aggregation. Omit atom metrics for all-padding chunks and weighted-force metrics when no positive real-atom preference remains.
- Preserve raw detail-file shapes/order. Spin force metrics now exclude padding as described in the follow-up below; the legacy spin-partner mapping, atomic-energy errors, and energy/virial per-atom normalization are unchanged.

The change is confined to the energy tester; shared training/full-validation metrics are not modified.

## Regression coverage

Controlled evaluator tests cover no padding, changing virtual-atom positions, unequal per-frame real-atom counts, zero/nonzero/NaN padding, all-padding chunks, fixed-type broadcasting, missing force labels, virtual-only preferences, chunk/system weighting, and detail-file layout.

The integration test writes and reloads a small **real PyTorch Hessian-enabled checkpoint**, creates an on-disk mixed-type dataset, and evaluates it with one-frame and multi-frame chunks. It independently checks MAE, RMSE, weights, and detail data for forces, weighted forces, and Hessians. It also runs `python -m deepmd --pt test` in a separate process. This is an initialized test model with real inference, not a mock evaluator or a trained scientific-potential benchmark.

For example, the integration dataset contains 7 real atoms in 9 padded slots: the force weight must be 21, not 27, and the Hessian weight 153, not 243.

## Validation

Built an independent checkout of upstream `58a12b1e0a72cfcc322c64d70c21469220bef74f` from source in WSL Ubuntu with Python 3.12.3, PyTorch 2.11.0+cpu, TensorFlow CPU 2.21.0 and real compiled operators; e3nn 0.6.0 is installed for the PyTorch model import path.

- Final new tests against unmodified production code: **14 failed, 5 passed**.
- With the fix, all **19 new tests passed**, and the complete focused/related run produced **56 passed**:

```sh
python -m pytest -q \
  source/tests/common/test_dp_test_virtual_atoms.py \
  source/tests/pt/test_dp_test_virtual_atoms.py \
  source/tests/common/test_dp_test_ener_split.py \
  source/tests/common/test_data_system_hessian.py \
  source/tests/pt/test_dp_test.py \
  source/tests/pt/test_weighted_avg.py \
  source/tests/tf/test_dp_test.py \
  source/tests/tf/test_virtual_type.py
```

Full-repository Ruff 0.16.0 lint and formatting, changed-file isort 9.0.0b1 and `git diff --check` passed. The three tested source blobs match the submitted files.

CPU-only validation; no GPU, JAX, full-project suite or complete pre-commit-hook run is claimed. The related suite emits TorchScript deprecation, CPU pin-memory, and multiprocessing-fork warnings. Upstream CI remains authoritative.

## Review follow-up: spin padding

Addressed the spin-force finding in `3f54a70`.

- The spin alignment helper can now exclude padding for metric inputs while its default still preserves the existing detail-file rows. Legacy TF spin-partner types remain magnetic degrees of freedom; they are not confused with type `-1` padding.
- The separate-magnetic-force layout intersects `mask_mag` with valid atom types. Both predictions and references are selected before subtraction, so padded NaNs cannot enter the statistics.
- Only nonempty, labeled real/magnetic selections produce metrics. The valid scalar counts feed the existing chunk/system weighted reduction; no training/full-validation metric code is changed.

New controlled tests exercise both spin layouts, no/changing/unequal/all-padding frames, fixed-type broadcasting, zero/nonzero/NaN padding on either or both inputs, missing labels, empty magnetic selections, detail-array preservation and nonmutation.

A second integration test saves/reloads a real initialized PyTorch spin checkpoint, evaluates on-disk mixed-type data with two chunk sizes, checks both real and magnetic force statistics and detail files, and runs `python -m deepmd --pt test` in a separate process. The dataset has **21 valid real-force scalars, not 27 padded scalars**, and 12 magnetic-force scalars.

Validation in the same independent native CPU build:

- Against the previous PR head's production blob `3e83d3f`: new controlled spin tests **82 failed / 22 passed**; new checkpoint tests **2 failed** (incorrect real-force count/MAE, or NaN contamination).
- With this follow-up: all **106 new tests passed**.
- Previous related suite plus the two new files: **162 passed**.
- Additional existing `source/tests/tf/test_deeppot_spin.py`: **4 passed**. Legacy padded metric selection is covered by controlled tests; this extra inference suite is not presented as a new padded-TF-checkpoint integration test.
- Full-repository Ruff lint/format, changed-file isort, velin at the repository's Git tag `0.0.12`, and `git diff --check` passed. The repository's E8001/E8002 checker also passed when explicitly registered with Pylint 4.0.6. The normal Pylint command emitted existing checker-registration warnings, so this is **not** a claim that every pre-commit hook ran successfully.
- Tested Linux and submitted Windows file blobs match. No GPU, JAX, full-project suite or trained-potential accuracy claim. Upstream CI/review remains pending.

Reproduce the added tests in a source-built CPU test environment:

```sh
python -m pytest -q \
  source/tests/common/test_dp_test_spin_virtual_atoms.py \
  source/tests/pt/test_dp_test_spin_virtual_atoms.py
```

## AI assistance

Implementation and local verification were prepared with OpenAI Codex. Local success is not maintainer approval or an upstream merge.

## Summary by CodeRabbit

- **Bug Fixes**
  - Force, Hessian, and weighted force metrics now exclude virtual or padding atoms.
  - Spin and magnetic-force metrics correctly exclude virtual atoms while preserving detail output.
  - Metrics remain consistent across mixed-type and fixed-type data, chunk sizes, and per-frame evaluation.
  - Datasets containing only virtual atoms no longer produce misleading metrics.
  - Force metrics are omitted when force labels are unavailable.
  - Padding represented by zeros or NaN values is handled correctly.

### CODE REVIEW

File: deepmd/infer/model_test/ener.py [hunk 1]
```diff
@@ -482,19 +487,33 @@ def evaluate_chunk(
             find_atom_pref=find_atom_pref,
         )
 
+        real_atoms = _real_atom_mask(atype, nframes, natoms)
+        has_real_atoms = bool(np.any(real_atoms))
+        metric_force = force
+        metric_reference_force = test_data["force"]
+        if self.reports_plain_force:
```
[the remark this answers is not in the document]
Comment (LOGO127) [reply deepmd/infer/model_test/ener.py:508]: Addressed in 3f54a70.

The spin path now selects non-padding metric inputs independently of the arrays returned for detail output. Legacy TF spin-partner types remain in the magnetic subset; the separate-magnetic-force layout also intersects mask_mag with valid atom types. Selection precedes subtraction on both predictions and references. Empty real/magnetic selections no longer emit zero-weight or NaN metrics, while missing magnetic arrays still raise when that label is requested.

Validation with the real production tester and source-built CPU operators:
- Previous PR production: new controlled tests 82 failed / 22 passed; two new real PyTorch spin-checkpoint integration cases failed on the padded real-force count/MAE or NaN contamination.
- With the follow-up: all 106 new cases pass; the combined related suite passes all 162 tests.
- Four additional existing TensorFlow spin-inference tests pass. Controlled tests cover legacy padded layout; I am not claiming a new padded-TF-checkpoint integration test.
- The PyTorch integration exercises disk data/checkpoint loading, single-/multi-frame chunks, raw real/magnetic detail files and a separate-process dp test command. 21 real-force scalars are counted instead of 27 padded scalars.
- Ruff, isort, velin and diff checks pass; the repository's custom dtype/device checker passes via explicit registration. The normal Pylint entry point emits checker-registration warnings; a complete pre-commit run is not claimed.

Codex assisted implementation and verification. Upstream CI and maintainer acceptance are still pending.

**Review by njzjz — APPROVED** [review]
